File: Neetcode Questions/Math & Geometry/happy-number.cpp

// https://leetcode.com/problems/happy-number/description/

class Solution {
public:

    int findSquare(int n){
        int ans = 0;
        while(n != 0){
            int lastDigit = n%10;
            ans += pow(lastDigit, 2);
            n = n/10;
        }

        return ans;
    }

    bool isHappy(int n) {
        
        int slow = n, fast = n;

        slow = findSquare(n);
        fast = findSquare(findSquare(n));

        while(slow != fast){
            slow = findSquare(slow);
            fast = findSquare(findSquare(fast));
        }

        if(slow == 1) return true;
        return false;
    }
};

File: Neetcode-150-js/Math & Geometry/happy-number.js

// https://leetcode.com/problems/happy-number/description/

/**
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function(n) {
    
    let slow = n;
    let fast = n;

    var findSquare = function(n){
        let ans = 0;
        while(n !== 0){
            let lastDigit = n%10;
            n = Math.floor(n/10);
            ans += Math.pow(lastDigit, 2)
        }
        return ans;
    }

    slow = findSquare(slow);
    fast = findSquare(findSquare(fast));

    while(slow !== fast){
        slow = findSquare(slow);
        fast = findSquare(findSquare(fast));
    }

    if(slow === 1) return true;
    return false;
};